modules/unsupported/process-feature/src/main/java/org/geotools/process/vector/FilteringVectorProcess.java

```diff
@@ -0,0 +1,423 @@
+                    // features with null values are skipped
+                    if (addGroupingValues(groupingValues, f))
+                        addDoubleValueFromFeature(f, valuesToCompare, beingFiltered);
+                } else {
+                    // is the feature in the group?
+                    if (featureComparison(groupingValues, f)) {
+                        addDoubleValueFromFeature(f, valuesToCompare, beingFiltered);
+                    } else {
+                        ((PushBackFeatureIterator) delegate).pushBack();
+                        break;
+                    }
+                }
+            }
+            next = doFiltering(beingFiltered, valuesToCompare);
+            return next != null;
+        }
+
+        private Feature doFiltering(
+                Map<Object, Feature> beingFiltered, List<Double> beingEvaluated) {
+            Object key;
+            // searches the min or max inside the values list and retrieve the feature
+            // to be returned from the Map.
+            if (beingFiltered.size() > 0 && beingEvaluated.size() > 0) {
+                if (this.aggregation.equals(Operations.MIN)) key = computeMin(beingEvaluated);
+                else key = computeMax(beingEvaluated);
+
+                return beingFiltered.get(key);
+            } else {
+                return null;
+            }
+        }
+
+        private Number computeMin(List<Double> beingEvaluated) {
+            return Collections.min(beingEvaluated);
+        }
+
+        private Number computeMax(List<Double> beingEvaluated) {
+            return Collections.max(beingEvaluated);
+        }
+
+        private boolean featureComparison(List<Object> groupingValues, Feature f) {
+            List<Object> toCompareValues = new ArrayList<>(groupingValues.size());
+            for (PropertyName p : groupByAttributes) {
+                toCompareValues.add(p.evaluate(f));
+            }
+            if (groupingValues.size() > 0 && groupingValues.equals(toCompareValues)) return true;
+            return false;
+        }
+
+        private boolean addGroupingValues(List<Object> groupingValues, Feature f) {
+            for (PropertyName p : groupByAttributes) {
+                Object result = p.evaluate(f);
+                if (result != null) groupingValues.add(p.evaluate(f));
+            }
+            if (groupingValues.isEmpty() || groupingValues.size() < groupByAttributes.size())
+                return false;
+            return true;
+        }
+
+        private void addDoubleValueFromFeature(
+                Feature f, List<Double> valuesToCompare, Map<Object, Feature> beingFiltered) {
+            Object result = this.operationAttribute.evaluate(f, Number.class);
+            if (result instanceof Number) {
+                Double value = ((Number) result).doubleValue();
+                valuesToCompare.add(value);
```

Comment:
Accumulating features in memory like this is dangerous, could lead to OOM if the groups are large, if all the operations needed are min and max, just keep track of the current value and current feature, and replace if the new feature is "better" than the old one. No need to keep the entire list. Then the method could be renamed to something like "updateSelection" or something like that.

Reply:
Thinking about it, the MinVisitor and MaxVisitor are already doing the right thing, make them visit the features one by one, get the result, reset them before visiting the next group, done.